### DISCUSSION TITLE
Support for enum(IntFlag) in addition to enum(IntEnum)?

### DISCUSSION THREAD

**Discussion by nurpax** [wjakob/nanobind#543] [Ideas]
Now that Nanobind has support for `enum(IntEnum)`, supporting `enum(IntFlag)` seems like a nice addition.

Here's a motivating case for it:

```
class InputTextFlags(enum.IntEnum):
    NONE = 0
    CHARS_DECIMAL = 1<<0
    CHARS_HEXADECIMAL = 1<<1
    UPPER = 1<<2

a1 = InputTextFlags.CHARS_HEXADECIMAL | InputTextFlags.UPPER # this type checks as 'a' becomes of type int
a: InputTextFlags = InputTextFlags.CHARS_HEXADECIMAL | InputTextFlags.UPPER # this doesn't
```

The latter case unfortunately doesn't type check:

![image](https://github.com/wjakob/nanobind/assets/297823/0156641f-f2f2-4881-8819-bf0b78438c62)

This can be fixed by typing functions differently, f.ex. by typing input flag arguments as `int` instead of `InputTextFlags`.

But Python enums support a more type safe alternative, the `enum.IntFlag.`

```
class InputTextFlags2(enum.IntFlag):
    NONE = 0
    CHARS_DECIMAL = 1 << 0
    CHARS_HEXADECIMAL = 1 << 1
    UPPER = 1 << 2

b: InputTextFlags2 = InputTextFlags2.CHARS_HEXADECIMAL | InputTextFlags2.UPPER
```

Perhaps there could be an enum annotation, say `nb::is_intflag()`, similar to the existing `nb::is_arithmetic()` flag to signal Nanobind to use `IntFlag` instead of `IntEnum`?

**Comment by wjakob** [comment]
I don't plan to work on this feature myself, but I am open to external contributions if they don't add too much extra complexity.
